ccm_web/server/src/canvas/canvas.service.ts

```diff
@@ -47,75 +51,136 @@ export class CanvasService {
+        `${this.url}/login/oauth2/token`, params
       ).toPromise()
       logger.debug(`Status code: ${response.status}`)
-      if (response.status === HttpStatus.OK) {
-        data = response.data
+      data = response.data
+    } catch (error) {
+      if (axios.isAxiosError(error) && error.response !== undefined) {
+        logger.error(`Received unusual status code ${error.response.status}`)
+        logger.error(`Response body: ${JSON.stringify(error.response.data, null, 2)}`)
       } else {
-        logger.error(`Received unusual status code ${response.status}`)
-        logger.error(`Response body: ${JSON.stringify(response.data, null, 2)}`)
+        logger.error(
+          `Error occurred while making request to Canvas for access token: ${JSON.stringify(error, null, 2)}`
+        )
       }
-    } catch (error) {
-      logger.error(
-        'Error occurred while making request to Canvas for access token: ',
-        JSON.stringify(error, null, 2)
-      )
+      throw new CanvasOAuthAPIError()
     }
-    if (data === undefined) return false
 
     // Create CanvasToken instance for user
-    let tokenCreated = false
+    const expiresAt = CanvasService.calculateExpiresAt(timeOfRequest, data.expires_in)
     try {
       await this.canvasTokenModel.create({
         userId: user.id,
         accessToken: data.access_token,
-        refreshToken: data.refresh_token
+        refreshToken: data.refresh_token,
+        expiresAt: expiresAt.toISOString()
       })
-      tokenCreated = true
       logger.info(`CanvasToken record successfully created for user ${userLoginId}.`)
     } catch (error) {
       logger.error(
-        'Error occurred while writing Canvas token data to the database: ',
-        JSON.stringify(error, null, 2)
+        `Error occurred while writing Canvas token data to the database: ${JSON.stringify(error, null, 2)}`
       )
+      throw new DatabaseError()
     }
-    return tokenCreated
   }
 
   async findToken (userLoginId: string): Promise<CanvasToken | null> {
     const user = await this.userService.findUserByLoginId(userLoginId)
-    if (user === null) throw new Error(`User with login ID ${userLoginId} was not found!`)
-
+    if (user === null) {
+      logger.error(`User ${userLoginId} is not in the database.`)
+      throw new UserNotFoundError(userLoginId)
+    }
     const token = user.canvasToken === undefined ? null : user.canvasToken
     return token
   }
 
+  async refreshToken (token: CanvasToken): Promise<CanvasToken> {
+    const params = {
+      grant_type: 'refresh_token',
+      client_id: this.clientId,
+      client_secret: this.secret,
+      refresh_token: token.refreshToken
+    }
+
+    let data: TokenRefreshResponseBody | undefined
+    const timeOfRequest = new Date()
+
+    try {
+      const response = await this.httpService.post<TokenRefreshResponseBody>(
+        `${this.url}/login/oauth2/token`, params
+      ).toPromise()
```

Comment:
what is the purpose of having `toPromise` here? this is something I noticed new.

Reply:
The `HttpService` pattern described here (https://docs.nestjs.com/techniques/http-module) returns an `rxjs` `Observable` (see https://rxjs.dev/guide/observable), which I don't really know how to use effectively. So I'm just calling this method to make it a `Promise` for now.

Reply:
Well, This is definitely unknown territory for me. So what ever works, we should go for it. 

Mostly the Observer pattern is used if 2 parts of the UI has same piece of information to update and these 2 UI modules/parts ( what ever you want to call) might be using 2 different service layers. So in that case these 2 parts will subscribe to same a source of truth and receive updates when the source of truth changes. For example, in the UI there might be 2 place where you have Thumbs up/down count and you want to keep both places in sync you will end up with Observable pattern.

With that said, I really don't know how that situation will comprehend to our scenario to receive token stuff. Probably what you did is correct.